bindings/java/src/main/native/jucx_common_def.cc

```diff
@@ -194,7 +194,9 @@ UCS_PROFILE_FUNC_VOID(jucx_request_callback, (request, status), void *request, u
 
 void recv_callback(void *request, ucs_status_t status, ucp_tag_recv_info_t *info)
 {
-    jucx_request_callback(request, status);
+    if (ucs_likely(UCS_PTR_IS_PTR(request))) {
+        jucx_request_callback(request, status);
```

Comment:
But request will always be a pointer in the callback. Also immediate completion is supposed to be a fast path (so likely case). Instead could you try check some filed in your request which is initialized when `ucp_tag_recv_nb` returns? If it is not set in the callback, the callback was called directly from `ucp_tag_recv_nb`. You will need to reset this request filed before calling `ucp_request_free`

Reply:
> But request will always be a pointer in the callback.
Here's what happen:
```
#0  0x00007ffff74bb9d7 in sched_yield () from /usr/lib64/libc.so.6
#1  0x00007fd8e2138d5b in jucx_request_callback(void*, ucs_status_t) () from /scrap/users/swat/jenkins/hadoop_tmp/nm-local-dir/usercache/swat/appcache/application_1572452711300_0004/container_1572452711300_0004_01_000015/tmp/jucx6872994377125644897/libjucx.so
#2  0x00007fd8e2138f2c in recv_callback(void*, ucs_status_t, ucp_tag_recv_info*) () from /scrap/users/swat/jenkins/hadoop_tmp/nm-local-dir/usercache/swat/appcache/application_1572452711300_0004/container_1572452711300_0004_01_000015/tmp/jucx6872994377125644897/libjucx.so
#3  0x00007fd8e238fe12 in ucp_request_complete_tag_recv (status=UCS_OK, req=0xb65700) at /hpc/scrap/users/swat/jenkins/ucx/contrib/../src/ucp/core/ucp_request.inl:109
#4  ucp_tag_request_process_recv_data (dereg=0, offset=<optimized out>, length=<optimized out>, data=<optimized out>, req=0xb65700) at /hpc/scrap/users/swat/jenkins/ucx/contrib/../src/ucp/tag/tag_match.inl:243
#5  ucp_tag_recv_request_process_rdesc (offset=<optimized out>, rdesc=0x7fd8df994e00, req=0xb65700) at /hpc/scrap/users/swat/jenkins/ucx/contrib/../src/ucp/tag/tag_match.inl:260
#6  ucp_tag_frag_list_process_queue (tm=tm@entry=0xc97c50, req=<optimized out>, msg_id=msg_id@entry=10568971807321856086, counter_idx=counter_idx@entry=3) at /hpc/scrap/users/swat/jenkins/ucx/contrib/../src/ucp/tag/tag_match.c:158
#7  0x00007fd8e239453a in ucp_tag_recv_common (debug_name=0x7fd8e23c1b8b "recv_nb", rdesc=0x7fd8df996000, cb=0x7fd8e2138f05 <recv_callback(void*, ucs_status_t, ucp_tag_recv_info*)>, req_flags=64, req=<optimized out>, tag_mask=<optimized out>, tag=<optimized out>, datatype=<optimized out>, count=<optimized out>, buffer=<optimized out>, worker=0xc97aa0) at /hpc/scrap/users/swat/jenkins/ucx/contrib/../src/ucp/tag/tag_recv.c:153
#8  ucp_tag_recv_nb_inner (cb=0x7fd8e2138f05 <recv_callback(void*, ucs_status_t, ucp_tag_recv_info*)>, tag_mask=<optimized out>, tag=<optimized out>, datatype=<optimized out>, count=<optimized out>, buffer=<optimized out>, worker=0xc97aa0) at /hpc/scrap/users/swat/jenkins/ucx/contrib/../src/ucp/tag/tag_recv.c:196
#9  ucp_tag_recv_nb (worker=0xc97aa0, buffer=<optimized out>, count=<optimized out>, datatype=<optimized out>, tag=<optimized out>, tag_mask=<optimized out>, cb=0x7fd8e2138f05 <recv_callback(void*, ucs_status_t, ucp_tag_recv_info*)>) at /hpc/scrap/users/swat/jenkins/ucx/contrib/../src/ucp/tag/tag_recv.c:179
#10 0x00007fd8e213ae36 in Java_org_openucx_jucx_ucp_UcpWorker_recvTaggedNonBlockingNative () from /scrap/users/swat/jenkins/hadoop_tmp/nm-local-dir/usercache/swat/appcache/application_1572452711300_0004/container_1572452711300_0004_01_000015/tmp/jucx6872994377125644897/libjucx.so
```

Inside `ucp_tag_recv_nb` it called callback with status `UCS_OK`.

Reply:
The issue is that it's possible to have progress in another thread, so that's why we handle case when we're in a callback, but the field in a request may not be initialized:

```
1. request = ucp_tag_recv_nb(..., callback); // progress in another thread
2. set_some_field_in_request(request)
```
For other operations, we guaranteed to not call callback if the operation completed immediately:
```
If the operation is @b not completed immediately and no error is
 * reported, then the UCP library will schedule invocation of the call-back
 * routine @a cb upon completion of the put operation.
```

For `ucp_tag_recv_nb` there's a case when it completes immediately and the callback is called.